File: epi_judge_python/lowest_common_ancestor_with_parent.py
import functools

from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

from binary_tree_with_parent_prototype import BinaryTreeNode


def lca(node0: BinaryTreeNode, node1: BinaryTreeNode):
    def get_depth(node):
        depth = 0
        while node:
            node = node.parent
            depth += 1
        return depth
    d0, d1 = get_depth(node0), get_depth(node1)
    if d0 > d1:
        diff = d0 - d1
        while diff > 0:
            node0 = node0.parent
            diff -= 1
    elif d1 > d0:
        diff = d1 - d0
        while diff > 0:
            node1 = node1.parent
            diff -= 1
    while node0 is not node1:
        node0 = node0.parent
        node1 = node1.parent

    return node0


@enable_executor_hook
def lca_wrapper(executor, tree, node0, node1):
    result = executor.run(
        functools.partial(lca, must_find_node(tree, node0),
                          must_find_node(tree, node1)))

    if result is None:
        raise TestFailure("Result can't be None")
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("lowest_common_ancestor_with_parent.py",
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))

File: epi_judge_cpp_solutions/lowest_common_ancestor_with_parent.cc
#include <algorithm>
#include <cstdlib>

#include "binary_tree_with_parent_prototype.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

using std::swap;

int GetDepth(const BinaryTreeNode<int>*);

BinaryTreeNode<int>* Lca(const unique_ptr<BinaryTreeNode<int>>& node0,
                         const unique_ptr<BinaryTreeNode<int>>& node1) {
  BinaryTreeNode<int>*iter0 = node0.get(), *iter1 = node1.get();
  int depth0 = GetDepth(iter0), depth1 = GetDepth(iter1);
  // Makes iter0 as the deeper node in order to simplify the code.
  if (depth1 > depth0) {
    swap(iter0, iter1);
  }
  // Ascends from the deeper node.
  int depth_diff = abs(depth0 - depth1);
  while (depth_diff--) {
    iter0 = iter0->parent;
  }

  // Now ascends both nodes until we reach the LCA.
  while (iter0 != iter1) {
    iter0 = iter0->parent, iter1 = iter1->parent;
  }
  return iter0;
}

int GetDepth(const BinaryTreeNode<int>* node) {
  int depth = 0;
  while (node->parent) {
    ++depth, node = node->parent;
  }
  return depth;
}

int LcaWrapper(TimedExecutor& executor,
               const unique_ptr<BinaryTreeNode<int>>& tree, int key0,
               int key1) {
  const unique_ptr<BinaryTreeNode<int>>& node0 = MustFindNode(tree, key0);
  const unique_ptr<BinaryTreeNode<int>>& node1 = MustFindNode(tree, key1);

  auto result = executor.Run([&] { return Lca(node0, node1); });

  if (!result) {
    throw TestFailure("Result can not be nullptr");
  }
  return result->data;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "tree", "key0", "key1"};
  return GenericTestMain(args, "lowest_common_ancestor_with_parent.cc",
                         "lowest_common_ancestor.tsv", &LcaWrapper,
                         DefaultComparator{}, param_names);
}
